### PR TITLE
Upgrade phpunit

### LINKED ISSUE #575: "mandatory" null validation is not working
See https://github.com/atk4/data/blob/1c5c361c2a8a34d0038fe9bf4e95f247b9f6997d/tests/FieldTest.php#L924
`b`/`mandatory` field should throw when null is set like the `c`/`required` field there.

As confirmed by @DarkSide666, the `set()` method is expected to fully normalize/validate the field at the time of setting the value (ie. immediatelly as `required` does).

Considering this as a Major as this is core functionality and many developers may rely on that.

### PR DESCRIPTION
**Pull request by mvorisek**
Also fix #575

### CODE REVIEW

**Comment by mvorisek** [comment]
Help wanted with the failing tests.

**Comment by mvorisek** [comment]
@DarkSide666 any ETA on this?

**Comment by mvorisek** [comment]
@DarkSide666 any ETA on this?

File: tests/ConditionTest.php [hunk 1]
```diff
@@ -16,7 +16,8 @@ public function testException1()
         // not existing field in condition
         $m = new Model();
         $m->addField('name');
-        $this->expectException(Exception::class);
+
+        $this->expectException(\atk4\core\Exception::class);
```
Comment (mvorisek) [inline tests/ConditionTest.php]: great!

File: tests/ReferenceSQLTest.php [hunk 2]
```diff
@@ -51,12 +51,12 @@ public function testBasic()
...
-        if ($this->driverType === 'sqlite') {
-            $this->assertSame(
-                'select "id","amount","user_id" from "order" where "user_id" in (select "id" from "user" where "id" > :a)',
-                $oo->action('select')->render()
-            );
-        }
+
+        $e = $this->getEscapeChar();
+        $this->assertSame(
+            str_replace('"', $e, 'select "id","amount","user_id" from "order" where "user_id" in (select "id" from "user" where "id" > :a)'),
```
Comment (mvorisek) [inline tests/ReferenceSQLTest.php:57]: better to put `$e` / expr. there directly

File: tests/ReferenceSQLTest.php [hunk 3]
```diff
@@ -69,12 +69,11 @@ public function testLink()
...
 
-        if ($this->driverType === 'sqlite') {
-            $this->assertSame(
-                'select "id","amount","user_id" from "order" where "user_id" = "user"."id"',
-                $u->refLink('Orders')->action('select')->render()
-            );
-        }
+        $e = $this->getEscapeChar();
+        $this->assertSame(
+            str_replace('"', $e, 'select "id","amount","user_id" from "order" where "user_id" = "user"."id"'),
```
Comment (mvorisek) [inline tests/ReferenceSQLTest.php:74]: the same here

File: tests/ReferenceSQLTest.php [hunk 4]
```diff
@@ -112,12 +111,11 @@ public function testLink2()
 
         $u->hasMany('cur', [$c, 'our_field' => 'currency_code', 'their_field' => 'code']);
 
-        if ($this->driverType === 'sqlite') {
-            $this->assertSame(
-                'select "id","code","name" from "currency" where "code" = "user"."currency_code"',
-                $u->refLink('cur')->action('select')->render()
-            );
-        }
+        $e = $this->getEscapeChar();
```
Comment (mvorisek) [inline tests/ReferenceSQLTest.php:114]: the same here

File: tests/ReferenceSQLTest.php [hunk 5]
```diff
@@ -154,12 +152,11 @@ public function testBasicOne()
...
         $o->addCondition('amount', '<', 9);
 
-        if ($this->driverType === 'sqlite') {
-            $this->assertSame(
-                'select "id","name" from "user" where "id" in (select "user_id" from "order" where "amount" > :a and "amount" < :b)',
-                $o->ref('user_id')->action('select')->render()
-            );
-        }
+        $e = $this->getEscapeChar();
+        $this->assertSame(
```
Comment (mvorisek) [inline tests/ReferenceSQLTest.php:156]: the same here

File: tests/ReferenceSQLTest.php [hunk 6]
```diff
@@ -228,12 +225,11 @@ public function testRelatedExpression()
...
         $i->addExpression('total_net', $i->refLink('line')->action('fx', ['sum', 'total_net']));
 
-        if ($this->driverType === 'sqlite') {
-            $this->assertSame(
-                'select "invoice"."id","invoice"."ref_no",(select sum("total_net") from "invoice_line" where "invoice_id" = "invoice"."id") "total_net" from "invoice"',
-                $i->action('select')->render()
-            );
-        }
+        $e = $this->getEscapeChar();
+        $this->assertSame(
```
Comment (mvorisek) [inline tests/ReferenceSQLTest.php:229]: the same here

File: tests/ReferenceSQLTest.php (hunk 2 above)
Comment (DarkSide666) [reply tests/ReferenceSQLTest.php:57]: That will make very long and unreadable SQL string.

File: tests/ConditionTest.php (hunk 1 above)
Comment (DarkSide666) [reply tests/ConditionTest.php]: addressed

**Review by DarkSide666 — APPROVED** [review]

**Pull request merged**